src/radiation/source_terms_multi_group.hpp

```diff
@@ -160,6 +160,89 @@ AMREX_GPU_DEVICE auto RadSystem<problem_t>::ComputeJacobianForGasAndDustDecouple
 	return result;
 }
 
+// Compute kappaF and the delta_nu_kappa_B_at_edge term. kappaF is used to compute the work term and the delta_nu_kappa_B_at_edge term is used to compute the
+// transport between groups in the momentum function. Only the last two arguments (kappaFVec, delta_nu_kappa_B_at_edge) are modified in this function.
+template <typename problem_t>
+AMREX_GPU_DEVICE void RadSystem<problem_t>::ComputeModelDependentKappaFAndDeltaTerms(
+    double const T, double const rho, amrex::GpuArray<double, nGroups_ + 1> const &rad_boundaries, quokka::valarray<double, nGroups_> const &fourPiBoverC,
+    quokka::valarray<double, nGroups_> const &kappaP, quokka::valarray<double, nGroups_> const &kappaE, quokka::valarray<double, nGroups_> &kappaF,
+    amrex::GpuArray<double, nGroups_> &delta_nu_kappa_B_at_edge)
+{
+	amrex::GpuArray<double, nGroups_> delta_nu_B_at_edge{};
+	const auto kappa_expo_and_lower_value = DefineOpacityExponentsAndLowerValues(rad_boundaries, rho, T);
+	for (int g = 0; g < nGroups_; ++g) {
+		auto const nu_L = rad_boundaries[g];
+		auto const nu_R = rad_boundaries[g + 1];
+		auto const B_L = PlanckFunction(nu_L, T); // 4 pi B(nu) / c
+		auto const B_R = PlanckFunction(nu_R, T); // 4 pi B(nu) / c
+		auto const kappa_L = kappa_expo_and_lower_value[1][g];
+		auto const kappa_R = kappa_L * std::pow(nu_R / nu_L, kappa_expo_and_lower_value[0][g]);
+		delta_nu_kappa_B_at_edge[g] = nu_R * kappa_R * B_R - nu_L * kappa_L * B_L;
+		delta_nu_B_at_edge[g] = nu_R * B_R - nu_L * B_L;
+	}
+	if constexpr (opacity_model_ == OpacityModel::piecewise_constant_opacity) {
+		kappaF = kappaP;
+	} else {
+		if constexpr (use_diffuse_flux_mean_opacity) {
+			kappaF = ComputeDiffusionFluxMeanOpacity(kappaP, kappaE, fourPiBoverC, delta_nu_kappa_B_at_edge, delta_nu_B_at_edge,
+								 kappa_expo_and_lower_value[0]);
+		} else {
+			// for simplicity, I assume kappaF = kappaE when opacity_model_ ==
+			// OpacityModel::PPL_opacity_full_spectrum, if !use_diffuse_flux_mean_opacity. We won't use this
+			// option anyway.
+			kappaF = kappaE;
+		}
+	}
+}
+
+// Compute kappaE and kappaP based on the opacity model. The result is stored in the last five arguments: alpha_B, alpha_E, kappaP, kappaE, and kappaPoverE.
+template <typename problem_t>
+AMREX_GPU_DEVICE void RadSystem<problem_t>::ComputeModelDependentKappaEAndKappaP(
+    double const T, double const rho, amrex::GpuArray<double, nGroups_ + 1> const &rad_boundaries, amrex::GpuArray<double, nGroups_> const &rad_boundary_ratios,
+    quokka::valarray<double, nGroups_> const &fourPiBoverC, quokka::valarray<double, nGroups_> const &Erad, int const n_iter,
+    amrex::GpuArray<double, nGroups_> &alpha_B, amrex::GpuArray<double, nGroups_> &alpha_E, quokka::valarray<double, nGroups_> &kappaP,
+    quokka::valarray<double, nGroups_> &kappaE, quokka::valarray<double, nGroups_> &kappaPoverE)
```

Comment:
This could also return a struct instead of having a large number of function arguments.

Reply:
Good point. Done.